### PR TITLE
Add Decode/Encode instances for Record, simplify module structure

### PR DESCRIPTION
**Pull request by paf31**
I left these instances out originally when I updated for 0.12 because I thought they wouldn't be necessary if we had `DecodeWithOptions` and `EncodeWithOptions`, but I've found some places in my code recently where they've been really useful.

Things have had to be moved around a bit to make the module order work out, but nothing much else has changed. I've taken the opportunity to simplify the module dependencies since this will be a breaking change anyway.

I'll leave this here for feedback for a few days before making a release.

### CODE REVIEW

**Comment by hdgarrood** [comment]
I think it might be preferable to explain the difference between the `Decode` and `Decode_` classes in doc-comments rather than in release notes, so that they show up on Pursuit. I think they're more likely to be seen that way.

**Comment by hdgarrood** [comment]
Other than that this looks great :+1:

File: generated-docs/Foreign/Class.md [hunk 1]
```diff
@@ -71,9 +87,75 @@ Encode Char
...
+```
+
+#### `DecodeWithOptions`
+
+``` purescript
+class DecodeWithOptions a  where
+  decodeWithOptions :: Options -> Foreign -> F a
+```
+
+When deriving `En`/`Decode` instances using `Generic`, we want
```
Comment (paf31) [inline generated-docs/Foreign/Class.md]: @hdgarrood How's this?

Comment (hdgarrood) [reply generated-docs/Foreign/Class.md]: Looks good! I particularly like the class rename

**Pull request merged**